benchmarks/src/main/kotlin/bfschannel/BfsChannelBenchmark.kt

```diff
@@ -0,0 +1,362 @@
+@file:JvmName("BfsChannelBenchmark")
+
+package bfschannel
+
+
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.channels.*
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.runBlocking
+import org.nield.kotlinstatistics.standardDeviation
+import java.io.FileInputStream
+import java.io.FileOutputStream
+import java.io.InputStreamReader
+import java.io.PrintWriter
+import java.net.URL
+import java.nio.channels.Channels
+import java.nio.file.Paths
+import java.util.*
+import java.util.concurrent.atomic.AtomicLong
+import java.util.zip.GZIPInputStream
+import kotlin.collections.ArrayList
+
+private val GRAPHS = listOf(
+        RandomGraphCreator("RAND-1M-10M", 1_000_000, 10_000_000),
+        DownloadingGraphCreator("USA-DISTANCE", GraphType.GR_GZ, "http://www.dis.uniroma1.it/challenge9/data/USA-road-d/USA-road-d.USA.gr.gz"),
+        DownloadingGraphCreator("LIVE-JOURNAL", GraphType.TXT_GZ, "https://snap.stanford.edu/data/soc-LiveJournal1.txt.gz"))
+/**
+ * Iterations count for each graph
+ */
+private const val ITERATIONS = 5
+/**
+ * Number of coroutines that are used to execute bfs in parallel
+ */
+private val COROUTINES = listOf(1, 4, 8, 16)
+/**
+ * Output file for the benchmark results
+ */
+private val RESULT_FILE = "plots-generators/resultsBfs.csv"
+
+/**
+ * This benchmark tests channel as a working queue, as a queue under contention.
+ * The benchmark creates or downloads graphs, then executes parallel BFS using channel as a queue, executes sequential BFS,
+ * compares the results and computes execution times.
+ *
+ * NB: this benchmark works painfully slow without syncronization on methods [kotlinx.coroutines.internal.LockFreeLinkedListNode.remove]
+ * and [kotlinx.coroutines.internal.LockFreeLinkedListNode.helpDelete] (or other channels fixes).
+ */
+fun main() {
+    val results = ArrayList<String>()
+    for (graphCreator in GRAPHS) {
+        val graphName = graphCreator.name
+        println("=== $graphName ===")
+        val graph = graphCreator.getGraph()
+
+        val startNode = graph[0]
+
+        // warmup
+        val sequentialResults = runSequentialBfs(startNode, graph)
+
+        // benchmark iterations
+        val sequentialExecutionTimes = (1..ITERATIONS).map { runSequentialBfs(startNode, graph).executionTime }
+
+        results += "$graphName,0,${sequentialExecutionTimes.average() / 1_000_000},${sequentialExecutionTimes.standardDeviation() / 1_000_000}"
+
+        println("sequential execution time = ${sequentialExecutionTimes.average() / 1_000_000}ms " +
+                "std = ${sequentialExecutionTimes.standardDeviation() / 1_000_000}ms")
+
+        for (coroutines in COROUTINES) {
```

Comment:
Here is a problem that we run both sequential and parallel code on the same JVM instance. Thus, it is possible that JIT re-optimizes the code during benchmarking.

Reply:
The main potential concern in creating a JVM instance for each parameters combination is consuming a lot of time for reading and parsing graphs. Could you please check whether this is a real problem?